src/historywork/CheckSingleLedgerHeaderWork.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2021 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "historywork/CheckSingleLedgerHeaderWork.h"
+#include "history/FileTransferInfo.h"
+#include "history/HistoryArchive.h"
+#include "history/HistoryManager.h"
+#include "historywork/GetAndUnzipRemoteFileWork.h"
+#include "util/Logging.h"
+#include "util/TmpDir.h"
+#include "xdrpp/printer.h"
+
+namespace stellar
+{
+
+CheckSingleLedgerHeaderWork::CheckSingleLedgerHeaderWork(
+    Application& app, std::shared_ptr<HistoryArchive> archive,
+    LedgerHeaderHistoryEntry const& expected)
+    : Work(app, fmt::format("check-single-ledger-header-{}",
+                            expected.header.ledgerSeq))
+    , mArchive(archive)
+    , mExpected(expected)
+{
+}
+
+CheckSingleLedgerHeaderWork::~CheckSingleLedgerHeaderWork()
+{
+}
+
+void
+CheckSingleLedgerHeaderWork::doReset()
+{
+    mGetLedgerFileWork.reset();
+    mDownloadDir =
+        std::make_unique<TmpDir>(mApp.getTmpDirManager().tmpDir(getName()));
+    uint32_t checkpoint = mApp.getHistoryManager().checkpointContainingLedger(
+        mExpected.header.ledgerSeq);
+    mFt = std::make_unique<FileTransferInfo>(
+        *mDownloadDir, HISTORY_FILE_TYPE_LEDGER, checkpoint);
+}
+
+BasicWork::State
+CheckSingleLedgerHeaderWork::doWork()
+{
+    if (!mGetLedgerFileWork)
+    {
+        CLOG_INFO(History, "Downloading ledger checkpoint {} from archive {}",
+                  mFt->baseName_gz(), mArchive->getName());
+        mGetLedgerFileWork = addWork<GetAndUnzipRemoteFileWork>(*mFt, mArchive);
+        return State::WORK_RUNNING;
+    }
+    else if (mGetLedgerFileWork->getState() != State::WORK_SUCCESS)
+    {
+        return mGetLedgerFileWork->getState();
```

Comment:
Silent error if download failed?

I guess it's expected if the node is on a ledger more recent than the latest checkpoint...
I wonder if we should merge this with `HistoryArchiveReportWork` as this would allow to first check if the archive is more recent than local node, and only perform the ledger header verification if the archive has it (and avoid bogus error messages like what this version does)

Reply:
Also, I don't think you want this `Work` to return failure *ever*:
* you don't want to retry
* you want to try the next archive in the list (it's possible that the failure is caused by a bad archive)

Reply:
I think you're right we should not be silent here, but wrong that we should never fail: we do want to retry (in case it's just an IO error or timing skew that's transient) and while I agree it might be nice to try the next archive in the list, the more important part is to reliably report an error (eg. have the `stellar-core self-check` command exit 1) when there is an error.

Updated code accordingly (returning error but also being noisy about it).

Reply:
> For the "online" version, we could wait (using consensus?) for some time past the next publish ledger?

Just wanted to add a note here that we do have `ConditionalWork` so we could add a condition to not start this work until the next checkpoint.

Reply:
Re first part: I've fixed this so that the online version now delays (using ConditionalWork, as @marta-lokhova suggests below) checking archives for checkpoints until LCL is greater than the checkpoint _after_ the next checkpoint. This gives it a full checkpoint-period to finish publishing, which should be adequate in reality. Offline version still checks immediately.

Re second part: I can't quite tell what you're asking for. Do you think I should be making a non-retrying but failure-absorbing work adaptor? Or a variant of WorkSequence that doesn't fail on first error?

Reply:
Ok fixed this now as follows:

  - Individual steps in sequence do not retry.
  - Sequence as a whole continues after first failure (still fails overall, but keeps going until last element).
  - Removed all old archive-access metrics (they were racy / non-actionable)
  - Added new archive-check success/failure metrics
  - Put on schedule so self-check runs automatically (default every 3 hours)